### ISSUE TITLE
Multiple outputs not supported

### ISSUE THREAD

**Issue by rwemay** [BHoM/Excel_UI#182]
#### Description:

Multiple outputs not supported.

#### How to replicate:

Use a method that has multiple outputs and only the last output will be returned. 

#### Expected behaviour:

Return a list of the outputs into the cell.

**Comment by rwemay** [comment]
@awakeman , just speaking to @alelom  on this, and think there are at least a couple of options: 
 - Option 1) return a list for every multi output. The user will need to select the list item in Excel. Will be a pain for any calculations running an execute command, but this might need to be the default behaviour. 
 - Option 2) return the first item in the output list by default - believe most workflows with push and execute, the user will want the objects/result, not the boolean/status. There would then need to be an option/way to get the status/boolean. 
 - Option 3) as option 2) but _only_ for push and execute, all other multiple outputs returning a list of the outputs that the user has to explode.

**Comment by rwemay** [comment]
@IsakNaslundBh see here FYI

**Comment by awakeman** [comment]
option 3 seems doable because we can just implement 2 but create a wrapper method for push in the excel toolkit that returns the list of objects.

**Comment by adecler** [comment]
I haven't looked in Excel yet but here is how I saw it working in my head back in the days: 
Assuming that a single output list is generated in Excel as a column of cells starting at the cell that contains the method call, then multiple output would simple be generated as a sequence of columns.

**Comment by rwemay** [comment]
Thanks @adecler , at the moment a single output returns a list object inside one cell (hence the greater need for Linq methods in Excel). I assume the rationale for doing it this way was to keep the list in the cell where the formula is. An alternative example is the explode method which creates an array in Excel of rows/columns of exploded properties and names. 

I've been trying to think of a 'have cake and eat it' method, and I think you're comment has helped. Couldn't we add some Excel specific methods that allow you to specify the output range, including whether to transpose? If no inputs, you would return an in-cell list. If inputs you'd get that plus the output values as simple values (no connection back to BHoM). 

Maybe worth a call on it soon, as I think some improvements would be good on Excel UI to help it scale. 

@awakeman , this PR works BTW, thanks!

**Comment by awakeman** [comment]
I have reservations doing it without being explicitly told by the user, when we fix #181 exploding the method inline will be trivial and keeps the user in control of the ranges that get consumed. Would also be good to add a transpose option to the explode too, i'll raise an issue for that.

**Comment by al-fisher** [comment]
As always 'have cake and eat it' !  🍰 😋 👍 


But mirroring the behaviour of auto-explode that will give a warning if you are accidentally expanding across cells with data already in them sounds very good actually

**Comment by awakeman** [comment]
Excel gives us the warning for free so if we are sure of the approach then i could implement that reasonably trivially. i'd rather not add a transpose option to every call though but we could probably add an engine method that you could wrap it in for transposition. can look at a prototype of that this week.

**Comment by rwemay** [comment]
Go for it!

**Comment by rwemay** [comment]
@al-fisher , @adecler , just to pick up on this. I don't think we should change the default behaviour for any of the methods to return results into multiple cells (other than the explode method) - either as an array or otherwise, for now. I think we should stick to the current default which is returning a list of objects into the same cell, and so for multiple outputs, this will be a list of lists. I've tested the version of this that provided that functionality and it works well. 

For returning results into multiple cells, which we definitely need to do, I think we need a more fundamental/UI level approach to this, not only to prevent unexpected cells from being written to, but also to give control over the range and transposition of the data being returned. This can either be by adding input arguments to every method, or (more likely?) creating a more robust/flexible explode method (or similarly named) that allows a human to specify a range for the output results (defaulting to the cell location), and transpose boolean. 

So in summary, recommending multiple outputs are as a list of lists for now.

**Comment by adecler** [comment]
@rwemay , I agree. One comment on your last sentence though

> So in summary, recommending multiple outputs are as a list of lists for now.

I think this should either be a dictionary or a custom object. There are a few benefits to that:
- We don't loose the name of the output. Say you have a method that filter a bunch of object (to be on theme with your current pet project) and that method return the valid object but also the remaining one as a second output (so basically the dispatch in GH). Without the name, the user cannot know for sure which one is which. 
- Custom objects and dictionaries can be exploded using Explode so the user can easily have the data spread across multiple cells if they choose to. Even with current functionality. 

On more comment on that second point: We could actually have an option in the context menu of a cell to explode its content on the spot or even a few menu items like `Explode vertically` and `Explode horizontally`. We can obviously still have the option to explode straight away as an extra param in the formula. But why not have both? 


**_Easter egg_**: I naively wanted to test if using `Explode` on a `BH.oM.Reflection.Output` was working properly in the UI and got this:

![image](https://user-images.githubusercontent.com/16853390/74000086-bf85ca00-49a1-11ea-8ea1-597ec08cace0.png)


So all working perfectly 😄

**Comment by rwemay** [comment]
@adecler this sounds fantastic and along the lines of our discussions offline. Shall we go for Dictionary output in single cell for multiple outputs as default? I think adding the explode arguments to the context menu with horizotally/vertically sub options sounds perfect. @al-fisher FYI
